File: 剑指Offer/04_replaceBlank.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char * replaceBlank(char * string);

int main(int argc, char const *argv[])
{
    char * string = (char *)malloc(sizeof(char)*50);
    // strcpy(string, "hello world");
    strcpy(string, "We are happy");
    printf("%s \n", string);
    printf("result: %s", replaceBlank(string));

    return 0;
}

char * replaceBlank(char * str) {
    char blank = ' ';
    int index = 0;
    int loop = 1;

    while (loop) {
        char temp = str[index];

        if (temp == blank) {
            int length = strlen(str);
            int j = length - 1;
            while (j > index) {
                str[j + 2] = str[j];
                j--;
            }
            str[index] = '%';
            str[index + 1] = '2';
            str[index + 2] = '0';
        }
        index++;
        if (index > strlen(str)) {
            loop = 0;
            break;
        }
    }
    
    return str;
}

File: 剑指Offer/04_replaceBlank.js
// Question 4 
/**
 * 请实现一个函数，把字符串中的每个空格替换成"%20"。
 * 例如输入“We are happy.”​，
 * 则输出“We%20are%20happy.”​。 
 */
function replaceBlank(obj) {
    const blank = ' ';
    let str = obj.str;

    let loop = true;
    let index = 0;

    while (loop) {
        let char = str[index];

        if (char === blank) {
            j = str.length - 1;
            while (j > index) {
                str[j + 2] = str[j];
                j--;
            }
            str[index] = '%';
            str[index + 1] = '2';
            str[index + 2] = '0';
        }
        index++;
        if (index > str.length) {
            loop = false;
            break;
        }

    }

    return str;
}
let string = 'hello world';
console.log(replaceBlank({str: 'hello world'}));



